src/util/cache/cache.c

```diff
@@ -5,75 +5,91 @@
  */
 
 #include "cache.h"
+#include "../../RG.h"
+#include <assert.h>
+#include <pthread.h>
+#include <stdatomic.h>
 #include "../rmalloc.h"
+#include "cache_array.h"
 
-//------------------------------------------------------------------------------
-// dict type callbacks
-//------------------------------------------------------------------------------
-
-// dict key compare
-int dictStringKeyCompare(void *privdata, const void *key1, const void *key2) {
-	return strcmp(key1, key2) == 0;
-}
-
-// dict key hash function
-uint64_t dictStringHash(const void *key) {
-	return HT_dictGenHashFunction(key, strlen(key));
-}
-
-// dict type
-dictType HT_dictTypeHeapStrings = {
-	dictStringHash,             // hash function
-	NULL,                       // key dup
-	NULL,                       // val dup
-	dictStringKeyCompare,       // key compare
-	NULL,                       // key destructor
-	NULL                        // val destructor
-};
-
-Cache *Cache_New(uint size, CacheItemFreeFunc freeCB) {
+Cache *Cache_New(uint cap, CacheEntryFreeFunc freeFunc) {
 	Cache *cache = rm_malloc(sizeof(Cache));
-	// Instantiate a new list to store cached values.
-	cache->list = CacheList_New(size, freeCB);
-	// Instantiate the lookup map for fast cache retrievals.
-	cache->lookup = HT_dictCreate(&HT_dictTypeHeapStrings, NULL);
+	cache->cap = cap;
+	cache->size = 0;
+	// Initialize counter to zero.
+	cache->counter = 0;
+	// Instantiate a new array to store cached values.
+	cache->arr = rm_calloc(cap, sizeof(CacheEntry));
+	cache->free_entry = freeFunc;
+	// Instantiate the lookup rax for fast cache retrievals.
+	cache->lookup = raxNew();
+	// Initialize the read-write lock to protect access to the cache.
+	assert(pthread_rwlock_init(&cache->_cache_rwlock, NULL) == 0);
 	return cache;
 }
 
-inline void *Cache_GetValue(const Cache *cache, const char *key) {
-	CacheListNode *elem = HT_dictFetchValue(cache->lookup, key);
-	if(elem == NULL) return NULL;
-
-	// Element is now the most recently used; promote it.
-	CacheList_Promote(cache->list, elem);
-	return elem->value;
+inline void *Cache_GetValue(Cache *cache, const char *key) {
+	pthread_rwlock_rdlock(&cache->_cache_rwlock);
+	CacheEntry *entry = raxFind(cache->lookup, (unsigned char *)key, strlen(key));
+	if(entry == raxNotFound) {
+		pthread_rwlock_unlock(&cache->_cache_rwlock);
+		return NULL;
+	}
+	// Element is now the most recently used; update its LRU.
+	// Multiple threads can be here simultaneously, make atomic updates.
+	ulong curr_counter = atomic_load(&cache->counter);
+	ulong curr_LRU = atomic_load(&entry->LRU);
+	atomic_compare_exchange_strong(&entry->LRU, &curr_LRU, curr_counter);
+	atomic_fetch_add(&cache->counter, 1);
```

Comment:
@alonre24 I'm not sure if this has any bearing on the non-determinism we observed before, but Clang complains about these lines with the error:
```
address argument to atomic operation must be a pointer to _Atomic type ('ulong *' (aka 'unsigned long *') invalid)
```